File: go/core/matcher.go
package varcore

import (
	"regexp"
	"sort"
)

// Matches a sentence against a registry's compiled expressions — port of
// matcher.ts / matcher.rs. Unanchored substring scan per step, then greedy
// left-to-right non-overlap resolution. All returned offsets are UTF-16 (regex
// byte offsets converted at Hit construction).

// paramSpan is the UTF-16 start/end of one captured parameter within the sentence.
type paramSpan struct {
	start int
	end   int
}

// hit is one successful expression match inside a sentence. formats aligns 1:1
// with args (nil where the parameter type has no formatter).
type hit struct {
	expression string
	stepDef    *StepRegistration
	matchStart int
	matchEnd   int
	args       []Value
	paramSpans []paramSpan
	formats    []FormatFn
}

// ambiguityCollision holds two or more hits that start at the same position
// with equal length.
type ambiguityCollision struct {
	matchStart int
	matchEnd   int
	candidates []hit
}

// resolvedSteps is the tagged result of resolveHits: when ambiguous is true,
// collisions holds every same-start/same-length tie; otherwise steps holds the
// greedy, left-to-right, non-overlapping selection.
type resolvedSteps struct {
	ambiguous  bool
	steps      []hit
	collisions []ambiguityCollision
}

// findHits returns every expression match found anywhere in sentence, one
// unanchored scan per registered step, in registration order. Regex byte
// offsets are converted to UTF-16 at Hit construction.
func findHits(sentence string, registry Registry) []hit {
	var hits []hit
	for _, step := range registry.Steps {
		unanchored, err := regexp.Compile(stripAnchors(step.compiled.regexpSource))
		if err != nil {
			continue
		}
		for _, loc := range unanchored.FindAllStringIndex(sentence, -1) {
			mStart, mEnd := loc[0], loc[1]
			matchedText := sentence[mStart:mEnd]
			arguments := step.compiled.matchWhole(matchedText)

			args := make([]Value, 0, len(arguments))
			var spans []paramSpan
			formats := make([]FormatFn, 0, len(arguments))
			for _, arg := range arguments {
				formats = append(formats, registry.Formats[arg.parameterTypeName])
				if arg.hasGroup {
					spans = append(spans, paramSpan{
						start: utf16Index(sentence, mStart+arg.groupStart),
						end:   utf16Index(sentence, mStart+arg.groupEnd),
					})
				}
				args = append(args, arg.value)
			}

			hits = append(hits, hit{
				expression: step.Expression,
				stepDef:    step,
				matchStart: utf16Index(sentence, mStart),
				matchEnd:   utf16Index(sentence, mEnd),
				args:       args,
				paramSpans: spans,
				formats:    formats,
			})
		}
	}
	return hits
}

// resolveHits selects the greedy, left-to-right, non-overlapping subset of hits,
// or reports every same-start/same-length ambiguity. Port of resolveHits.
func resolveHits(hits []hit) resolvedSteps {
	if len(hits) == 0 {
		return resolvedSteps{steps: nil}
	}
	sorted := make([]hit, len(hits))
	copy(sorted, hits)
	// Sort by matchStart ascending, then by length descending (stable).
	sort.SliceStable(sorted, func(i, j int) bool {
		if sorted[i].matchStart != sorted[j].matchStart {
			return sorted[i].matchStart < sorted[j].matchStart
		}
		return (sorted[j].matchEnd - sorted[j].matchStart) < (sorted[i].matchEnd - sorted[i].matchStart)
	})

	var collisions []ambiguityCollision
	i := 0
	for i < len(sorted) {
		hereStart := sorted[i].matchStart
		hereLen := sorted[i].matchEnd - sorted[i].matchStart
		j := i + 1
		for j < len(sorted) &&
			sorted[j].matchStart == hereStart &&
			sorted[j].matchEnd-sorted[j].matchStart == hereLen {
			j++
		}
		if j-i > 1 {
			cands := make([]hit, j-i)
			copy(cands, sorted[i:j])
			collisions = append(collisions, ambiguityCollision{
				matchStart: hereStart,
				matchEnd:   sorted[i].matchEnd,
				candidates: cands,
			})
		}
		i = j
	}
	if len(collisions) > 0 {
		return resolvedSteps{ambiguous: true, collisions: collisions}
	}

	var steps []hit
	cursor := -1
	for _, h := range sorted {
		if h.matchStart < cursor {
			continue
		}
		cursor = h.matchEnd
		steps = append(steps, h)
	}
	return resolvedSteps{steps: steps}
}

File: rust/core/src/matcher.rs
//! Matches a sentence against a registry's compiled expressions — port of
//! `matcher.ts` / `Matcher.java`. Unanchored substring scan per step, then
//! greedy left-to-right non-overlap resolution. All returned offsets are UTF-16
//! (regex byte offsets converted at [`Hit`] construction).

use crate::offsets::utf16_index;
use crate::registry::{FormatFn, Registry, StepRegistration};
use crate::value::Value;
use regex::Regex;
use std::rc::Rc;

/// UTF-16 start/end of one captured parameter within the sentence.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct ParamSpan {
    pub start: usize,
    pub end: usize,
}

/// One successful expression match inside a sentence. `formats` aligns 1:1 with
/// `args` (`None` where the parameter type has no formatter).
#[derive(Clone)]
pub struct Hit {
    pub expression: String,
    pub step_def: Rc<StepRegistration>,
    pub match_start: usize,
    pub match_end: usize,
    pub args: Vec<Value>,
    pub param_spans: Vec<ParamSpan>,
    pub formats: Vec<Option<FormatFn>>,
}

/// Two or more hits that start at the same position with equal length.
#[derive(Clone)]
pub struct AmbiguityCollision {
    pub match_start: usize,
    pub match_end: usize,
    pub candidates: Vec<Hit>,
}

/// The tagged result of [`resolve_hits`].
pub enum ResolvedSteps {
    /// The greedy, left-to-right, non-overlapping selection.
    Ok(Vec<Hit>),
    /// Every same-start/same-length tie that blocked selection.
    Ambiguous(Vec<AmbiguityCollision>),
}

/// Every expression match found anywhere in `sentence`, one unanchored scan per
/// registered step, in registration order. Port of `findHits`. Regex byte offsets
/// are converted to UTF-16 at [`Hit`] construction.
pub fn find_hits(sentence: &str, registry: &Registry) -> Vec<Hit> {
    let mut hits = Vec::new();
    for step in &registry.steps {
        let Ok(unanchored) = Regex::new(&strip_anchors(step.compiled.regexp_source())) else {
            continue;
        };
        for m in unanchored.find_iter(sentence) {
            let matched_text = &sentence[m.start()..m.end()];
            let arguments = step.compiled.match_whole(matched_text).unwrap_or_default();

            let mut args = Vec::with_capacity(arguments.len());
            let mut param_spans = Vec::new();
            let mut formats = Vec::with_capacity(arguments.len());
            for arg in arguments {
                formats.push(registry.formats.get(&arg.parameter_type_name).cloned());
                if let Some((gs, ge)) = arg.group {
                    param_spans.push(ParamSpan {
                        start: utf16_index(sentence, m.start() + gs),
                        end: utf16_index(sentence, m.start() + ge),
                    });
                }
                args.push(arg.value);
            }

            hits.push(Hit {
                expression: step.expression.clone(),
                step_def: step.clone(),
                match_start: utf16_index(sentence, m.start()),
                match_end: utf16_index(sentence, m.end()),
                args,
                param_spans,
                formats,
            });
        }
    }
    hits
}

/// Strips a compiled expression's `^...$` anchors so an unanchored scan can find
/// it anywhere in the sentence.
fn strip_anchors(source: &str) -> String {
    let s = source.strip_prefix('^').unwrap_or(source);
    let s = s.strip_suffix('$').unwrap_or(s);
    s.to_string()
}

/// Selects the greedy, left-to-right, non-overlapping subset of `hits`, or
/// reports every same-start/same-length ambiguity. Port of `resolveHits`.
pub fn resolve_hits(hits: Vec<Hit>) -> ResolvedSteps {
    if hits.is_empty() {
        return ResolvedSteps::Ok(Vec::new());
    }
    let mut sorted = hits;
    // Sort by matchStart ascending, then by length descending (stable).
    sorted.sort_by(|a, b| {
        a.match_start
            .cmp(&b.match_start)
            .then_with(|| (b.match_end - b.match_start).cmp(&(a.match_end - a.match_start)))
    });

    let mut collisions = Vec::new();
    let mut i = 0;
    while i < sorted.len() {
        let here_start = sorted[i].match_start;
        let here_len = sorted[i].match_end - sorted[i].match_start;
        let mut j = i + 1;
        while j < sorted.len()
            && sorted[j].match_start == here_start
            && sorted[j].match_end - sorted[j].match_start == here_len
        {
            j += 1;
        }
        if j - i > 1 {
            collisions.push(AmbiguityCollision {
                match_start: here_start,
                match_end: sorted[i].match_end,
                candidates: sorted[i..j].to_vec(),
            });
        }
        i = j;
    }
    if !collisions.is_empty() {
        return ResolvedSteps::Ambiguous(collisions);
    }

    let mut steps = Vec::new();
    let mut cursor: isize = -1;
    for hit in sorted {
        if (hit.match_start as isize) < cursor {
            continue;
        }
        cursor = hit.match_end as isize;
        steps.push(hit);
    }
    ResolvedSteps::Ok(steps)
}
